### PR TITLE
Assessment Configurability in Ground Control

### PR DESCRIPTION
**Pull request by jayjay19630**
### Description

This pull request aims to allow users to configure on an assessment basis. This will allow finer control of each individual assessment's features, such as token counters and voting features.
 
Screenshot 2024-03-22 at 4 30 02 PM

Additionally, the admin panel will have general assessment configuration toggles for token counters and voting features. However, toggles such as Token Counter and Voting Features will only activate these features upon the creation of an assessment. Any further changes involving token counters and voting features must be done within ground control.

Screenshot 2024-03-22 at 4 30 17 PM

### Type of change

- [x] New feature (non-breaking change which adds functionality)

### Checklist
- [x] I have tested this code

### CODE REVIEW

**Comment by DesSnowy** [comment]
Corresponding backend PR: https://github.com/source-academy/backend/pull/1093

**Comment by RichDom2185** [comment]
@joeng03 @martin-henz if you get to reviewing this before I do, please don't immediately merge it upon approval as I expect this to be a 3-way merge conflict with 2 other PRs.

**Review by joeng03 — APPROVED** [review]
LGTM

File: src/pages/academy/groundControl/subcomponents/GroundControlConfigureCell.tsx [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    [hasTokenCounter]
+  );
+  const toggleVotingFeatures = useCallback(
+    () => setHasVotingFeatures(!hasVotingFeatures),
+    [hasVotingFeatures]
+  );
+  const toggleIsTeamAssessment = useCallback(
+    () => setIsTeamAssessment(!isTeamAssessment),
+    [isTeamAssessment]
+  );
```
Comment (RichDom2185) [inline src/pages/academy/groundControl/subcomponents/GroundControlConfigureCell.tsx:52]: Nit: Here, we memoize, but we recreate every time `isXXXX` changes; this is not very ideal.

When toggling booleans, if we use the callback notation for `setXXX`, we can simply create the function once and not have to recreate it again:

```suggestion
  const toggleHasTokenCounter = useCallback(
    () => setHasTokenCounter(prev => !prev),
    []
  );
  const toggleVotingFeatures = useCallback(
    () => setHasVotingFeatures(prev => !prev),
    []
  );
  const toggleIsTeamAssessment = useCallback(
    () => setIsTeamAssessment(prev => !prev),
    []
  );
```

File: src/pages/academy/groundControl/subcomponents/GroundControlConfigureCell.tsx [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        <DialogBody>
+          <p>
+            This <b>configuration tool</b> allows you to fine-tune this assessment. Any changes made
+            here will <b>override</b> any assessment configurations in the admin panel.
+          </p>
+          <div className="general-configs">
+            <p>
+              <b>General Configurations</b>
+            </p>
+            <Divider></Divider>
```
Comment (RichDom2185) [inline src/pages/academy/groundControl/subcomponents/GroundControlConfigureCell.tsx:73]: Nit:

```suggestion
            <Divider />
```

File: src/pages/academy/groundControl/subcomponents/GroundControlConfigureCell.tsx [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+            <Switch
+              className="has-token-counter"
+              checked={hasTokenCounter}
+              onChange={toggleHasTokenCounter}
+              inline
+              label="Enable token counter"
+            ></Switch>
```
Comment (RichDom2185) [inline src/pages/academy/groundControl/subcomponents/GroundControlConfigureCell.tsx:80]: Nit:

```suggestion
            />
```

File: src/pages/academy/groundControl/subcomponents/GroundControlConfigureCell.tsx (hunk 2 above)
Comment (RichDom2185) [reply src/pages/academy/groundControl/subcomponents/GroundControlConfigureCell.tsx:73]: Ditto for the other occurrences in this file.

File: src/pages/academy/groundControl/subcomponents/GroundControlConfigureCell.tsx [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+                  inline
+                  label="Publish Voting (Coming soon!)"
+                ></Switch>
+              </div>
+            </Collapse>
+          </div>
+        </DialogBody>
+        <DialogFooter
+          actions={
+            <>
```
Comment (RichDom2185) [inline src/pages/academy/groundControl/subcomponents/GroundControlConfigureCell.tsx:141]: Unnecessary fragment.

**Review by RichDom2185 — COMMENTED** [review]
Hi, thanks for working on this, looking good! I just have some minor comments below.

As discussed, the order of merging will be:

* PR #2548
* PR #2875
* This PR

**Comment by jayjay19630** [comment]
I've polished up the code to fix the above comments. Pls let me know if there are any further issues!

**Review by RichDom2185 — APPROVED** [review]
LGTM, thanks a lot!

**Pull request merged**